fb-adb.1
.\" Copyright (c) 2014, Facebook, Inc.
.\" All rights reserved.
.\"
.\" This source code is licensed under the BSD-style license found in
.\" the LICENSE file in the root directory of this source tree. An
.\" additional grant of patent rights can be found in the PATENTS file
.\" in the same directory.
.\"
.TH fb-adb 1 2014-12-22 Facebook fb-adb
.SH NAME
fb-adb \- Improved Android Debug Bridge
.SH SYNOPSIS

.de rcmdargs
.OP \-dehrtTU
.OP \-E exename
.OP \-H host
.OP \-P port
.OP \-p product
.OP \-s device
.OP \-u user
.OP \-C dir
..

.SY "fb-adb shell"
.rcmdargs
.RI [ command ]
.RI [ args
.IR .\|.\|. ]
.YS

.SY "fb-adb rcmd"
.rcmdargs
.I command
.RI [ args
.IR .\|.\|. ]
.YS

.SY "fb-adb logw"
.OP \-der
.OP \-T tag
.OP \-x priority
.OP \-H host
.OP \-P port
.OP \-p product
.OP \-s device
.OP \-u user
.I message
.YS

.SY "fb-adb readlink"
.OP \-der
.OP \-H host
.OP \-P port
.OP \-p product
.OP \-s device
.OP \-u user
.I link
.YS

.SY "fb-adb help"
.YS

.SY "fb-adb"
.I adb-command
.RI [ args
.IR .\|.\|. ]
.YS

.SH DESCRIPTION

\fBfb-adb\fR wraps \fBadb\fR to provide enhanced shell and command
execution functionality.

\fBfb-adb shell\fR (which can also be spelled \fBfb-adb sh\fR and
\fBfb-adb shellx\fR) runs \fIcommand\fR and its arguments through the
device shell (which is usually \fBmksh\fR(1)) with all arguments
shell-quoted.  (\fIcommand\fR itself is unquoted, so it may contain
shell constructs, e.g., globs.)
.B fb-adb shell
without arguments starts an interactive shell, making it slightly more
efficient, especially when connecting to CPU-constrained devices.

.B fb-adb rcmd
is similar, except that it runs \fIcommand\fR directly without
invoking the shell and does not have a zero-argument form that starts
an interactive shell.

.B fb-adb help
displays a mixture of \fBfb-adb\fR's own help and \fBadb\fR's.  If
standard output is a terminal, the help is shown in a pager
(\fBPAGER\fR, which defaults to \fBless\fR).

All other commands are forwarded to \fBadb\fR, making it feasible to
use \fBfb-adb\fR in place of \fBadb\fR for most tasks.

\fBadb\fR must be on PATH.

.SH DIFFERENCES FROM ADB

Differences between \fBadb\fR and \fBfb-adb\fR are that \fBfb-adb\fR:

  * is binary clean (no LF -> CRLF mangling)
  * transmits and updates terminal window size via SIGWINCH
  * distinguishes standard output and standard error
  * properly muxes streams, each with independent flow control
  * allows for ssh-like pty allocation control
  * propagates program exit status instead of always exiting with status 0
  * properly escapes program arguments
  * kills remote program on local exit
  * provides a generic facility to elevate to root without re-escaping command
line arguments
  * transparently supports alternate, faster transports for shell connections
  * supports sending environment variables to the child out-of-band

.SH OPTIONS

.TP
.BR \-d
Connect to a device.  Identical to \fBadb\fR's \fB\-d\fR option.

.TP
.BR \-e
Connect to an emulator.  Identical to \fBadb\fR's \fB\-e\fR option.

.TP
.BI \-s " specific-device"
Connect to a specific device.  Identical to \fBadb\fR's \fB\-s\fR
option.

.TP
.BI \-p " product-name-or-path"
Supplied the product name or path.  Identical to \fBadb\fR's \fB\-p\fR
option.

.TP
.BI \-H " hostname"
Specifies the hostname of the \fBadb\fR server to which to connect.
Identical to \fBadb\fR's \fB\-H\fR option.

.TP
.BI \-P " hostname"
Specifies the port of the \fBadb\fR server to which to connect.
Identical to \fBadb\fR's \fB\-P\fR option.

.TP
.BR \-r ", " --root
Attempt to become root before executing \fIcommand\fR or starting an
interactive shell.

.TP
.BI \-u " user"
.TQ
.BI \--user " user"
Become \fIuser\fR before executing \fIcommand\fR or starting an
interactive shell.  \fIuser\fR is an identifier that \fBrun-as\fR on
the device understands, e.g., \fBcom.facebook.wakizashi\fR.

.TP
.BR \-U ", " \--socket
Use an AF_UNIX socketpair for the remote child's standard input and
standard output.

.TP
.BR \-D ", " \--no-ctty
Do not give the remote child a controlling terminal.  Normally,
\fBfb-adb\fR connects a remote child to a controlling terminal so that
when the a \fBfb-adb\fR disconnects, the child receives a SIGHUP.
With this option set, the child instead sees its standard streams
closed, but is not sent any signal.

.TP
.BR \-t ", " \--force-tty
Normally, if a \fIcommand\fR is supplied, \fBfb-adb\fR gives the
remote child pipes for its standard streams.  This option instructs
\fBfb-adb\fR to instead use a pseudoterminal for each standard stream
connected to a terminal on the client.  If repeated, \fB-t\fR makes
\fBfb-adb\fR give the child a pseudoterminal even for streams for
which the \fBfb-adb\fR client does not have a terminal.

.TP
.BR \-T ", " \--disable-tty
Never give the child a pseudo-terminal for a standard stream.

.TP
.BI \-E " exename"
.TQ
.BI \--exename " exename"
Execute \fIexename\fR as the remote child program instead of
\fIcommand\fB.  \fIcommand\fR is argv[0] in any case.  If starting an
interactive shell, use \fIcommand\fR instead of the default shell
executable.

.TP
.BI \-C " dir"
.TQ
.BI \--chdir " dir"
Change to directory \fIdir\fR before running \fIcommand\fR or starting
an interactive shell.

.TP
.BI \-X " transport"
.TQ
.BI \--transport " transport"
Connect over the given transport.  The default is \fBshell\fR.

\fBshell\fR means to run the \fBfb-adb\fR protocol over an
"adb shell" connection.  This method is the most reliable.

\fBunix\fR means to run the \fBfb-adb\fR protocol over an AF_UNIX
socket connection forwarded over ADB.  This method may transfer data
faster than the \fBshell\fR transport, but it is experimental may take
slightly longer to set up.

\fBtcp:HOST,PORT\fR means to use a TCP socket to connect from the
device to the host.  fb-adb automatically manages making the
connection, but it cannot automatically figure out on what network
address the device can reach the host.  You must tell it what address
to bind to: HOST and PORT.  HOST and PORT are any strings accepted as
the node and service arguments, respectively, to getaddr(3), and can
by names or numbers.

  Examples: "tcp:10.2.5.1,12345" and "tcp:myhost.example,gopher"

This transport is typically much faster than the others: it
bypasses adb completely.

Note that traffic on this connection is unencrypted.

.TP
.BI \-Y " NAME=VALUE"
.TQ
.BI \--setenv " NAME=VALUE"
Set the environment variable NAME to VALUE in the child.

\fBfb-adb\fR begins with the default environment inherited from the
shell.  \fBfb-adb\fR then applies the environment modifications
described by \fB-Y\fR, \fB-K\fR, and \fB-F\fR in order.  The child
inherits the resulting environment.

.TP
.BI \-K " NAME"
.TQ
.BI \--unsetenv " NAME=VALUE"
Unset the environment variable NAME in the child.

.TP
.BI \-F
.TQ
.BI \--clearenv
Unset all environment variables in the child.

.TP
.BR \-h ", " \--help ", " help
Display help output.

.SH ENVIRONMENT VARIABLES

.TP
.BI ADB_SHELL_OLD_BEHAVIOR
When set, \fBfb-adb\fR does not override any \fBadb\fR commands,
exposing only new names for new commands.  Using this option,
\fBfb-adb\fR can replace \fBadb\fR (perhaps on PATH) with no
compatibility risk.

.TP
.BI FB_ADB_DEBUG
This variable controls how \fBfb-adb\fR writes debug messages.  If it
is "1", \fBfb-adb\fR writes debug messages to standard error.  If it
is a string that begins with ">", \fBfb-adb\fR writes debug messages
to the file whose names begins after the ">" (like a shell
redirection).  If it is a string that begins with ">>", \fBfb-adb\fR
appends to that file instead of overwriting it.  This environment
variable only has an effect when \fBfb-adb\fR is compiled with the
--enable-checking option.

.TP
.BI FB_ADB_REMOTE_DEBUG
This option, set on the host system, becomes the value of
\fBFB_ADB_DEBUG\fR on the target system.

.TP
.BI FB_ADB_REMOTE_WRAPPER
This option, set on the host system, provides a program that wraps the
execution of the stub on the remote system.  This environment variable
has an effect only when the program is configured with
--enable-checking.  An example:

FB_ADB_REMOTE_WRAPPER='strace -o/data/local/tmp/foo'

.TP
.BI FB_ADB_NO_COMPRESSION
This option, set on the host system, tells \fBfb-adb\fR to avoid its
usual LZ4 stream compression.  LZ4 is fast enough that there are no
practical disadvantages to disabling it, so this option is primarily
useful for debugging.

.TP
.BI FB_ADB_TRANSPORT
This environment variable provides the default value of the
--transport option.

.TP
.BI TEMP
.TQ
.BI TMP
.TQ
.BI TMPDIR
When \fBfb-adb\fR needs to create a temporary file, it uses the first
of these directories that is set.  It defaults to /tmp if none of
these variables is present.

.SH LOGW COMMAND

\fBfb-adb logw\fR is a utility to write messages to logcat
on device.  The message logged is MESSAGE.

.TP
.BI \-t " TAG"
.TQ
.BI \--tag " TAG"
Log mesage with the given tag instead of the default, "fb-adb-logw".

.TP
.BI \-x " PRIORITY"
.TQ
.BI \--priority " PRIORITY"
Set logcat priority to PRIORITY.  PRIORITY is a case-insensitive
substring of "verbose", "debug", "informational",
"warning", "error", or "fatal".

.TP
.BR \-r ", " --root
Attempt to become root for the operation.

.TP
.BI \-u " user"
.TQ
.BI \--user " user"
Become \fIuser\fR for the operation.
interactive shell.  \fIuser\fR is an identifier that \fBrun-as\fR on
the device understands, e.g., \fBcom.facebook.wakizashi\fR.

.SH READLINK COMMAND

\fBfb-adb readlink\fR is a utility to read the values of symbolic
links on device.  It is useful when devices lack the readlink(1)
command.  It prints the given symbolic link's value to standard output
(with no trailing newline) or fails.

.TP
.BR \-r ", " --root
Attempt to become root for the operation.

.TP
.BI \-u " user"
.TQ
.BI \--user " user"
Become \fIuser\fR for the operation.
interactive shell.  \fIuser\fR is an identifier that \fBrun-as\fR on
the device understands, e.g., \fBcom.facebook.wakizashi\fR.

.SH BUGS

\fBfb-adb\fR should present better error messages when it can't run a
stub.  It should not try to send stubs for architectures that cannot
possibly work.

SELinux policy prevents the "socket" transport option from working
correctly under Android Lollipop or later.  See Android issue 1151.

.SH EXAMPLES

.IP "Capture a screenshot from device and write it locally to a timestamped file"
.B fb-adb rcmd screencap -p > screenshot-$(timestamp).png

.IP "Dump database.db of the com.bar.foo app"
.B fb-adb rcmd -u com.bar.foo sqlite3 /data/data/com.bar.foo/databases/database.db .d

.IP "Open remote shell as the user com.bar.foo"
.B fb-adb shell -u com.bar.foo

.SH SEE ALSO

Run
.B fb-adb help
to see ADB's own list of commands.


NEWS
Version 1.5.1
-------------

New features:

 * --user and --root options for logw and readlink.

Bugfixes:

 * Properly report errors executing run-as and su.

Version 1.5.0
-------------

New features:

 * readlink command.  Useful on older devices without a readlink(1).

Version 1.4.4
-------------

Bug fixes:

  * Make SIGINT and other signals work for delegated adb tasks.

Version 1.4.3
-------------

Bug fixes:

  * Fix kqueue-based ppoll emulation on OS X to work correctly with
    regular disk files.

Version 1.4.2
-------------

Bug fixes:

 * Abstract unix socket creation function should fail on OS X

Version 1.4.1
-------------

New features:

 * Faster ADB-forwarded and TCP-based transit modes
 * Environment variables for setting options
 * logw command for writing logcat entries on device
   (think of logger(1)).
 * LZ4 compression of traffic
 * Tuned buffer sizes for greater throughput
 * Set and unset environment variables in remote child environments

Bug fixes:

 * Build system now works properly in cross-compilation environment
 * Stub is never world-writable while on device
 * Clean up reliably on quit signals
 * Avoid deadlock during hard shutdown when partial msg in buffer
 * Safe concurrent stub uploading
 * More errors propagated from stub to parent in human-readable form
 * SELinux workarounds

Version 1.2
-----------

  * Initial stable release


fb-adb.c
/*
 *  Copyright (c) 2014, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in
 *  the LICENSE file in the root directory of this source tree. An
 *  additional grant of patent rights can be found in the PATENTS file
 *  in the same directory.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <stdint.h>
#include <assert.h>
#include <errno.h>
#include <string.h>
#include <sys/wait.h>
#include "util.h"
#include "strutil.h"
#include "cmd_shex.h"
#include "cmd_stub.h"
#include "cmd_logw.h"
#include "cmd_readlink.h"
#include "timestamp.h"

#ifndef __ANDROID__
static void
view_with_pager(const char* filename)
{
    system(xaprintf("%s %s",
                    getenv("PAGER") ?: "less",
                    xshellquote(filename)));
}
#endif

static void
usage(void)
{
    FILE* out = stdout;
#ifndef __ANDROID__
    const char* tmpf_name = NULL;
    if (isatty(1)) {
        out = xnamed_tempfile(&tmpf_name);
        allow_inherit(fileno(out));
    }
#endif

    fprintf(out, "%s %s - Enhanced ADB\n", prgname, PACKAGE_VERSION);
    fprintf(out, "\n");
    fprintf(out, "  %s shell [CMD ARGS...] - Improved adb shell.\n", prgname);
    fprintf(out, "\n");
    fprintf(out, "  %s rcmd CMD [ARGS...] - Run commands directly, "
            "without\n", prgname);
    fprintf(out, "    using the shell.\n");
    fprintf(out, "\n");
    fprintf(out, "  Use %s rcmd -h for additional help.\n", prgname);
    fprintf(out, "  Other commands forward to adb. See below.\n");
    fprintf(out, "\n");
    fflush(out);
    system(xaprintf("adb help >& %d 2>&1", fileno(out)));

#ifndef __ANDROID__
    if (tmpf_name) {
        view_with_pager(tmpf_name);
    }
#endif
}

#define DECLARE_FORWARDER(name)                                 \
    int name##_wrapper_main(int argc, const char** argv)        \
    {                                                           \
        return shex_wrapper(                                    \
            #name,                                              \
            name##_opts,                                        \
            name##_longopts,                                    \
            name##_usage,                                       \
            argv);                                              \
    }

DECLARE_FORWARDER(logw);
DECLARE_FORWARDER(readlink);

static bool
word_follows_adb_arg_p(const char* p)
{
    for (; *p; ++p) {
        if (strchr("spHP", *p)) {
            if (*(p+1) == '\0')
                return true;

            return false;
        }
    }

    return false;
}

static bool
adb_arg_p(char** argv, unsigned i)
{
    if (argv[i] == NULL)
        return false;

    if (argv[i][0] == '-' && argv[i][1] == 'h')
        return false;

    if (argv[i][0] == '-' &&
        argv[i][1] == '-' &&
        argv[i][2] != '\0')
    {
        return false;
    }

    if (i > 1 && !strcmp(argv[i-1], "--"))
        return false;

    if (argv[i][0] == '-')
        return true;

    /* Catch FOO in pairs like -s FOO */
    if (i > 1 &&
        argv[i-1][0] == '-' &&
        word_follows_adb_arg_p(argv[i-1]))
    {
        return true;
    }

    return false;
}

int
real_main(int argc, char** argv)
{
    unsigned nr_adb_args = 0;
    while (adb_arg_p(argv, nr_adb_args + 1))
        nr_adb_args += 1;

    unsigned non_adb_off = 1 + nr_adb_args;
    char* prgarg = argv[non_adb_off];
    if (prgarg == NULL)
        die(EINVAL, "no sub-command given. Use --help for help.");

    int (*sub_main)(int, const char**) = NULL;
    if (!strcmp(prgarg, "stub")) {
        sub_main = stub_main;
#if FBADB_MAIN
    } else if (!strcmp(prgarg, "shellx") || !strcmp(prgarg, "sh")) {
        sub_main = shex_main;
    } else if (!strcmp(prgarg, "shell") &&
               !getenv("ADB_SHELL_OLD_BEHAVIOR"))
    {
        sub_main = shex_main;
    } else if (!strcmp(prgarg, "rcmd")) {
        sub_main = shex_main_rcmd;
#endif
    } else if (!strcmp(prgarg, "logw")) {
#if FBADB_MAIN
        sub_main = logw_wrapper_main;
#else
        sub_main = logw_main;
#endif
    } else if (!strcmp(prgarg, "readlink")) {
#if FBADB_MAIN
        sub_main = readlink_wrapper_main;
#else
        sub_main = readlink_main;
#endif
    } else if (!strcmp(prgarg, "help") ||
               !strcmp(prgarg, "-h") ||
               !strcmp(prgarg, "--help"))
    {
        usage();
        return 0;
    } else if (!strcmp(prgarg, "--version")) {
#ifdef HAVE_GIT_STAMP
        printf("fb-adb %s git:%s\n", PACKAGE_VERSION, git_stamp);
#else
        printf("fb-adb %s\n", PACKAGE_VERSION);
#endif
        return 0;
    } else {
        sigprocmask(SIG_SETMASK, &orig_sigmask, NULL);
        execvp("adb", argv);
        die(EINVAL, "could not exec adb: %s", strerror(errno));
    }

    char* new_argv0 = xaprintf("%s %s", prgname, prgarg);
    set_prgname(new_argv0);
    argv[0] = new_argv0;
    memmove(&argv[non_adb_off],
            &argv[non_adb_off+1],
            sizeof (*argv) * (argc - nr_adb_args - 1));
    return sub_main(argc - 1, (const char**) argv);
}
